File: buildTreeFromIn_Post.py
'''
Time Complexity -->

O(n)

Space Complexity -->
O(n^2) since we are creating array multiple times
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if inorder:
            index_val = inorder.index(postorder.pop(-1))
            root = TreeNode(inorder[index_val])
            root.right = self.buildTree(inorder[index_val+1:], postorder)
            root.left = self.buildTree(inorder[:index_val], postorder)
            return root
        

File: BuildTreeFromInPost.java
// Time Complexity : O(n) //build HashMap, recursion
// Space Complexity : O(n) // HashMap + recursive stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : -
import java.util.HashMap;
import java.util.Map;

/*
Same as building tree from Inorder and Preorder but, for postorder the root will be at the end of the array and when we iterate backwards we get Right then Left
So, helper function needs to eval root, right,left resp.
 */
public class BuildTreeFromInPost {
    int pIdx;//postorder iterative index- starts from len-1
    Map<Integer,Integer> inMap = new HashMap<>(); //Map to store indexes of inorder elements
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        int idx = 0;
        pIdx = postorder.length -1;

        inMap = new HashMap<>();
        for(int i : inorder)
            inMap.put(i,idx++);

        return helper(postorder,0,inorder.length-1);
    }

    private TreeNode helper(int[] postorder, int sIdx, int eIdx){

        //base
        if(sIdx>eIdx) return null;

        //root value
        int rootVal = postorder[pIdx];
        pIdx--;

        TreeNode root = new TreeNode(rootVal);

        int rIdx = inMap.get(rootVal);
        root.right = helper(postorder, rIdx + 1, eIdx);
        root.left = helper(postorder, sIdx, rIdx -1);

        return root;
    }

    public static void main(String[] args) {
        BuildTreeFromInPost p = new BuildTreeFromInPost();
        TreeNode n = p.buildTree(new int[]{9,3,15,20,7}, new int[]{9,15,7,20,3});
    }
}
